0000-richer-documents.rst

```diff
@@ -0,0 +1,300 @@
+with that type. Most users will quite likely then want to use ``>>=`` to process
+those annotations and render them using GHC's default rendition or a custom
+one:::
+
+    (>>=) :: SDoc a -> (a -> SDoc b) -> SDoc b
+
+The ``Monad`` is exactly what is required to substitute all annotations by some
+annotation-free document that depends on the annotation value. In our scenario,
+``b`` will be ``()`` or ``Void``, indicating the "annotation-free" nature.
+We could even imagine tweaking the functions that actually print documents to
+only accept annotation-free documents, so as to force users to interpret
+annotations one way or another before getting the documents printed somewhere.
+
+In fact, GHC uses ``SDoc`` in other contexts than error messages, a major
+one being code generation. In those cases, we will never want to emit
+any annotation, just pure text, and as fast as possible. This would be another
+case where we would want to deal with ``SDoc Void`` values. It is therefore
+desired that any potential implementation of this proposal doesn't drastically
+change the performance document construction and rendering when no annotation is
+involved.
+
+Pieces of code that don't produce annotations in documents can either
+produce documents that are completely polymorphic in the annotation type,
+or ``SDoc Void`` values. The polymorphic variant has the merit of being
+usable in code that produces documents with a specific annotation type. This
+would let us print snippets of expressions inside error messages
+(``SDoc ErrAnnotation``), or in the output generated by some ``-ddump-*`` flag
+(``SDoc Void``).
+
+However, with the changes described so far, we would run into a problem. While
+the semantics of the existing ``Doc`` constructors are pretty clear in terms of
+layout, there is no good answer when interpreting a ``Pure a``. We will never
+know for sure what text this annotation is going to end up being replaced with,
+since the point of this proposal exactly consists in sprinkling non-textual
+Haskell values all over our documents and rendering them later. This in turns
+means that we can't reliably "guess" whether our annotation is going to end up
+being rendered over several lines, nor how many columns or levels of nesting it
+will involve.
+
+To work around this problem, we suggest to adopt the trick used in the
+`wl-pprint-extras <https://hackage.haskell.org/package/wl-pprint-extras>`_
+library, which consists in introducing constructors that allow users to
+introduce (sub-)documents that are dependent on the exact column number,
+nesting level, ribbon length, etc. For instance, this is how we would define
+the constructor that "suspends" a sub-document on the column number of the
+current position in the textual rendering of a larger document:::
+
+    data Doc a
+      = ...
+      | Pure a
+      | Column (Int -> Doc a) -- new!
+      | ...
+
+While this now prevents ``Doc`` from being ``Foldable`` or ``Traversable``
```

Comment:
Wait -- I'm confused. The text above said that we could support `Foldable` and `Traversable`. Here it says we can't. Which is it? I do think that `Foldable` would be useful: it would allow an IDE to e.g. extract all the types eagerly from an error message and then do some processing -- without rendering the message. I don't think `Foldable` is a must-have, if it's hard to support, but I do think it would be useful.

Actually, without a `Foldable` instance (or similar), how would an IDE extract the information it wants? I guess it would have to pattern-match on the `Doc` type directly.

Or, we would have a function

```hs
extractAnnotations :: SDocContext -> SDoc a -> [AnnotationInContext a]
extractAnnotations = ...

data AnnotationInContext a
  = AIC { annot :: a, annotPrintedRow :: Int, annotPrintedColBegin :: Int, annotPrintedColEnd :: Int }
```

or some such. This way, the tool gets all the annotations with the information it needs to syntax-highlight etc. the right bits of the output. Of course, the tool could also ignore the row/col information if appropriate.

Maybe this was all clear to you, but I think this could be spelled out more in the proposal.

Reply:
Your suggestion is quite interesting in its own right, but for now I'll just clarify my reasoning/idea and then we can decide which way we want the proposal to go.

The basic idea is that when we extend our existing `Doc` with that new `| Pure a` constructor, what we have _is_ `Foldable` and `Traversable`. At that point, we can use `toList` to collect annotations and `Traversable` to applicatively transform them. But when we add our extra constructors for introducing sub-documents that depend on the current column number, nesting level and what not, we lose those instances. The statement about having `Foldable` and `Traversable` only applies to the version of `Doc` that the proposal has at that point of the document, not the final version with the extra few constructors. I can look into making this clearer.

Note that `Foldable` and `Traversable` were already out of the window for `SDoc`, right from the start. So the idea I had was simply to offer a function like `collect :: SDoc a -> [a]`, where `collect` assumes that the document doesn't make use of the `Column`/`Nesting`/... constructors, and clearly states so.

As far as I can tell, our inspiration for this approach -- [wl-pprint-extras](http://hackage.haskell.org/package/wl-pprint-extras-3.5.0.5/docs/Text-PrettyPrint-Free.html#t:SimpleDoc) -- gets away by defining a `SimpleDoc` type, that doesn't have those `Column`/etc constructors nor some more typical document constructors that `Doc` has, and maps `Doc`s to `SimpleDoc`s, which are then displayed. Interestingly, `SimpleDoc` still supports annotations, but the scoped kind: `| Effect e (SimpleDoc e)`.

Reply:
Ah -- this helps. But actually, I think we will want to be able to output the annotations with location information, as IDEs will often want that. So maybe both functions are warranted: one that can get the annotations directly, and one that requires e.g. ribbon width to get them but comes with output location info.

Incidentally, if we have `collect :: SDoc a -> [a]`, then surely we can write a `Foldable` instance. Maybe it's best not to, if we're worried about invariants being broken. But perhaps we expose an interface that does not allow end users (that is, all of GHC outside the `Outputable` module) to make location-dependent annotations. Then, the `Foldable` instance would be safe to use.

Reply:
Well, what would we do in the clause for the `| Column (Int -> Doc a)` constructor of `Doc`?

``` hs
collect :: Doc a -> [a]
...
collect (Column k) = collect (k 0)
...
```

?

That would not be right if `k` is, say, `\i -> if i /= 0 then Pure i else Empty`.

Reply:
If the public (that is, outside the `Outputable` module) interface to the pretty-printer does not allow the construction of an annotation hidden behind a condition dependent on a column number, then we can safely do just what you wrote above. That is, given a `Doc` `Column f`, it is an invariant that the result of `collect (f n)` does not depend on `n`.

Reply:
Your "if" is something that we will have to check while implementing this proposal, then, I suppose. It is hard to take this conversation further without more details about where/when those constructors need to be introduced or eliminated. In case the assumption doesn't hold, there isn't much we can do without a rendered document; any annotation that's only available for a given column number argument or something along those lines is a symptom of a bad document. Those column number/etc arguments should only be relevant for positioning the elements relatively to one another, not for determining their contents. So yes, actually, maybe we can just "cheat" here, and pass dummy column/etc numbers when we want to collect the annotations.

Reply:
Perhaps I've meant my "if" as: let's try to have it be so. Because we get nice properties as a result!

Reply:
Yes, it's not unreasonable to think we can succeed. I'll revise the document accordingly, to outline our reasoning around this property.